Fix/message log viewer

danimarinBG opened the pull request:
Description

This pull request introduces HTML rendering support in the QgsMessageLogViewer, allowing developers and plugin authors to display rich-text log messages (bold, colored, or formatted text) directly in the QGIS “Log Messages” panel.

Currently, all messages passed to QgsMessageLog are HTML-escaped before being rendered, which prevents any kind of rich formatting — even when the message content is intentionally crafted as HTML.
This limitation affects many existing plugins that rely on visually distinguishable log messages to improve usability, readability, and user experience.

With this change:

HTML content sent through QgsMessageLog.logMessage() is now rendered correctly in the viewer (using appendHtml() with unescaped text).

Plain-text messages continue to work exactly as before.

The default behavior remains safe, since plugin authors still need to explicitly include HTML content in their messages.

Motivation

Many QGIS plugins (for example, Giswater, Plugin Reloader, and others) use formatted messages (e.g., colored tags, bold text, or links) to enhance user feedback in the Log panel.
Enabling native support for HTML makes this experience more consistent, avoids the need for custom message bars or dialogs, and helps plugin developers maintain a clear and readable log output for end users.

Implementation notes

The modification is minimal and localized to QgsMessageLogViewer::logMessage().

It removes the HTML escaping step (toHtmlEscaped()), allowing appendHtml() to render the intended formatting.

The change has been tested with both plain text and HTML content to ensure backward compatibility.

No API changes, no behavioral regressions, and no dependencies introduced.

<img width="1105" height="467" alt="image" src="https://github.com/user-attachments/assets/b9bbb7b6-4e63-4ac8-a362-b522672d5b6c" />

danimarinBG committed:
Add MessageType enum and update logMessage method to support message types

Introduced a new MessageType enum to differentiate between text and HTML messages in the application. Updated the logMessage method in QgsMessageLogViewer to accept a MessageType parameter, allowing for appropriate handling of message formatting based on the type.

danimarinBG committed:
Refactor message handling to use Qgis::MessageType enum

Updated QgsMessageOutput, QgsMessageOutputConsole, and QgsMessageViewer classes to utilize the Qgis::MessageType enum for message type handling. This change enhances type safety and consistency across message output methods.

danimarinBG committed:
Update logMessage method signature in QgsMessageLogViewer to use msgType parameter

Modified the logMessage method to replace the 'type' parameter with 'msgType', providing a default value for improved clarity and consistency in message logging.

danimarinBG commented:
As shown in the image below, log messages can now include rich formatting such as bold text, colors, indentation, and even links directly within the Log panel.
<img width="1609" height="378" alt="image" src="https://github.com/user-attachments/assets/13959de9-6545-4504-868e-c97aa15719ee" />
This greatly improves readability and user experience, especially for plugins like Giswater, Plugin Reloader, and many others that rely on visually structured feedback for users.

nyalldawson commented on src/core/qgis.h:
```diff
@@ -161,6 +161,17 @@ class CORE_EXPORT Qgis
+      MessageText, //!< Text message
+      MessageHtml, //!< HTML message
+    };
+    Q_ENUM( MessageType )
```
To explain:

- We always use enum class for newly introduced enums
- Convention is to avoid repeating the enum name in the enum values, hence dropping "Message" from "MessageText" (since enum classes always require the enum name, it helps keep things shorter without dropping the readability)
- MessageType is a little too generic => LogMessageType
- Drop the "this will be used..." text, as that kind of explanation goes out of date very quickly in QGIS code!
- The "SIP..." macros will cause Python compatibility code to be generated so that we don't break existing code/API by moving and renaming the enum

nyalldawson commented on src/gui/qgsmessagelogviewer.h:
```diff
@@ -48,7 +48,7 @@ class GUI_EXPORT QgsMessageLogViewer : public QDialog, private Ui::QgsMessageLog
     /**
      * Logs a \a message to the viewer.
```
```suggestion
     * Logs a \a message to the viewer.
     *
     * The \a type argument was added in QGIS 4.0
```

nyalldawson commented on src/gui/qgsmessagelogviewer.cpp:
```diff
@@ -169,7 +169,15 @@ void QgsMessageLogViewer::logMessage( const QString &message, const QString &tag
                            .arg( color.name(), QDateTime::currentDateTime().toString( Qt::ISODate ), levelString );
-  QString cleanedMessage = message.toHtmlEscaped();
+  
+  if ( msgType == Qgis::MessageType::MessageHtml )
```
use a switch here

danimarinBG committed:
Refactor enum naming and parameter conventions

- Rename MessageType to LogMessageType for better specificity
- Rename enum values: MessageText -> Text, MessageHtml -> Html
- Replace abbreviated parameter names (msgType -> type)
- Add SIP compatibility macros for Python API backward compatibility
- Use switch statements instead of if-else for enum handling

Addresses review comments from @nyalldawson

danimarinBG commented:
Hello, I see that the build is failing because there are still references in the C++ code to `QgsMessageOutput::MessageHtml`.
I'm not sure if I should replace these references with the new enum `Qgis::LogMessageType`, since the name now includes the word Log.
These references occur in several files: `qgisapp.cpp`, `qgsappcoordinateoperationhandlers.cpp`, `qgsrunprocess.cpp`, `qgsmodeldesignerdialog.cpp`, `qgsmodelgraphicsscene.cpp`, and `qgspythonutilsimpl.cpp`.

Could you confirm whether all these should be updated to `Qgis::LogMessageType::Html` (or the appropriate enum value) or if there are specific cases where the old MessageHtml should be retained?

danimarinBG committed:
Update message handling to use Qgis::StringFormat enum

Refactored message output methods across various components to utilize the new Qgis::StringFormat enum, replacing the previous LogMessageType. This change enhances consistency in message formatting, allowing for better handling of plain text and HTML messages throughout the application.

nyalldawson commented on src/core/qgis.h:
```diff
@@ -161,6 +161,18 @@ class CORE_EXPORT Qgis
+    /**
+     * \brief Format of strings
+     *
+     * \since QGIS 4.0
```
```suggestion
     * \since QGIS 4.0. Prior to QGIS 4.0 this was available as QgsMessageOutput.MessageType
```

danimarinBG committed:
Enhance message logging to include Qgis::StringFormat parameter

Updated the logMessage methods across QgsMessageLog, QgsMessageLogConsole, and QgsMessageLogViewer to accept a Qgis::StringFormat parameter.

danimarinBG committed:
Update QgsProcessingExec to include Qgis::StringFormat in message logging

Modified the connection to QgsMessageLog::messageReceived to maintaining consistency with recent changes in message handling.

nyalldawson commented on src/core/qgsmessagelog.h:
```diff
@@ -71,7 +71,7 @@ class CORE_EXPORT QgsMessageLog : public QObject
      * value for the message.
      */
-    void messageReceived( const QString &message, const QString &tag, Qgis::MessageLevel level );
+    void messageReceived( const QString &message, const QString &tag, Qgis::MessageLevel level, Qgis::StringFormat type );
```
Unfortunately I missed this in the earlier review, but adding the new argument to this signal breaks stable API.

We'll need to add a new signal with the new signature, so that existing connections in plugins continue to work.

Maybe something like `messageReceivedWithType`. 

Then whenever we emit the existing messageReceived signal, we'll also emit the new messageReceivedWithType signal. 

Optionally the old signal could be marked as deprecated with the Q_DECL_DEPRECATED / SIP_DEPRECATED macros

danimarinBG committed:
Refactor QgsMessageLog to introduce messageReceivedWithType signal

Updated the message logging system to include a new signal, messageReceivedWithType, which allows for the emission of messages along with their type. Adjusted connections in QgsMessageLogConsole, QgsMessageLogViewer, and QgsProcessingExec to utilize this new signal, enhancing message handling consistency across the application.

danimarinBG committed:
Fix compilation: remove deprecated messageReceived signal emission

- Remove emit of deprecated messageReceived signal from emitMessage()
- Regenerate SIP files for messageReceivedWithType
- Update Python tests to use messageReceivedWithType

Fixes -Werror=deprecated-declarations build failure.

danimarinBG commented:
Hi @nyalldawson,

The macOS build passed because the deprecated warning didn't stop the build, while Ubuntu/Fedora failed with -Werror. 

Should we make macOS fail on deprecated usage too for consistency?

I've also updated the Python tests to use messageReceivedWithType and regenerated the SIP files.

danimarinBG committed:
Fix Windows build: add missing qgis.h include in qgsmessageoutput.h

qgsmessageoutput.h uses Qgis::StringFormat but was missing qgis.h include,
causing Windows compilation errors due to stricter include order enforcement.

danimarinBG committed:
Merge branch 'master' into fix/message-log-viewer

danimarinBG commented:
Hi @nyalldawson, this PR was closed prematurely and the work was nearly complete.
Would it be possible to reopen it, or should I open a new PR referencing this one?

danimarinBG committed:
fix(format): run pre-commit command

danimarinBG commented:
I’m currently unsure how to fix the reported issues locally.
Running the sip commands regenerates many files that seem unrelated to this PR, so I’m not sure which of them (if any) should be committed.
Any guidance on the correct local workflow for this would be greatly appreciated.
